presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java

```diff
@@ -344,17 +407,17 @@ else if (parameterType.isPrimitive() && !hasNullFlag) {
                         checkArgument(!nullableArgument, "Method [%s] has parameter with primitive type %s annotated with @SqlNullable", method, parameterType.getSimpleName());
                     }
 
-                    if (typeParameterNames.contains(type.value()) && !(parameterType == Object.class && nullableArgument)) {
-                        // Infer specialization on this type parameter. We don't do this for @SqlNullable Object because it could match a type like BIGINT
-                        Class<?> specialization = specializedTypeParameters.get(type.value());
-                        Class<?> nativeParameterType = Primitives.unwrap(parameterType);
-                        checkArgument(specialization == null || specialization.equals(nativeParameterType), "Method [%s] type %s has conflicting specializations %s and %s", method, type.value(), specialization, nativeParameterType);
-                        specializedTypeParameters.put(type.value(), nativeParameterType);
+                    inferSpecialization(method, parameterType, type, nullableArgument);
+                    if (isBlockAndPositionNullConvention) {
+                        argumentNativeContainerTypes.add(type.nativeContainerType());
+                    }
+                    else {
+                        assertEquals(type.nativeContainerType(), Object.class);
```

Comment:
assertEquals is for assertion in test only. Here, you can use `checkArgument`. In addition, you would want to provide an error messages. The audience of the error message is people who are implementing custom SQL functions through the annotation message.